### PR TITLE
Advanced Websocket Features

### PR DESCRIPTION
**Pull request by Archmonger**
### New Features
- [x] Automatic WS reconnection (`settings.py:IDOM_WS_RECONNECT_TIMEOUT`)
- [x] Propogate WS down to components
- [x] Support authenticated components (Login the user and save the session if auth middleware is detected)
- [x] Pin `"idom-client-react": "^0.33.3"`
- [x] Update tests to add `websocket` as a component parameter
- [x] Update tests to use `render` shortcut instead of `HttpResponse`
- [x] Change `ALLOWED_HOSTS = ["*"]` in tests to allow LAN access 
- [x] Rename `IdomAsyncWebSocketConsumer` -> `IdomAsyncWebsocketConsumer`
- [x] Readme clean up. fixes, and updates

### CODE REVIEW

**Comment by Archmonger** [comment]
I decided it would be best to propagate the entire websocket down, to fall in line with Django view convention of propagating down the HTTP request.

A benefit of this is that the developer would now be able to utilize any WS functionality (`scope`, `close()`, etc) from within their component.

If we need to add any features, can be added as props to the `IdomAsyncWebSocketConsumer`

**Comment by Archmonger** [comment]
if I don't complete this PR by Thursday, I likely won't be available to pick it back up until October 4th.

**Comment by Archmonger** [comment]
Reconnecting WS has been developed.

Currently works on client side disconnect.

### There are two caveats...
1) Reconnect is not graceful.
   - All IDOM modules are effectively "reset" to the first rendered state upon reconnect.
   - All variable values are lost.
   - We may need to implement a database store for linking a WS session to component state to resolve this.
2) Reconnections after server restart is broken due to Django IDOM design.
   - We only add modules to `IDOM_REGISTERED_COMPONENTS` upon template tag render, so all components are completely unknown after server restart (`"Unknown IDOM view ID"`).
   - We will need to develop the [previously mentioned `IDOM_COLLECT_OFFLINE`](https://github.com/idom-team/django-idom/issues/4) feature to resolve this.

**Comment by Archmonger** [comment]
If approved, the last limb of Conreq blocking factors is `react-router` support.

Unfortunately I'm too much of a ReactJS scrub to PR that into IDOM Core, so I'll need your support on that.

File: src/django_idom/websocket_consumer.py [hunk 1]
```diff
@@ -15,6 +18,12 @@
 _logger = logging.getLogger(__name__)
 
 
+@dataclass
+class WebSocketConnection:
+    scope: dict
+    disconnect: Callable
```
Comment (rmorshea) [inline src/django_idom/websocket_consumer.py:24]: Can you make this a bit more explicit? Is this supposed to be `Callable[[], None]`?

File: src/js/src/index.js [hunk 2]
```diff
@@ -4,28 +4,34 @@ import { mountLayoutWithWebSocket } from "idom-client-react";
 let LOCATION = window.location;
 let WS_PROTOCOL = "";
 if (LOCATION.protocol == "https:") {
-  WS_PROTOCOL = "wss://";
+	WS_PROTOCOL = "wss://";
```
Comment (rmorshea) [inline src/js/src/index.js:7]: It looks like there's a lot of unnecessary changes here. Can you make the indent here two spaces rather than three?

**Review by rmorshea — CHANGES_REQUESTED** [review]
Besides those few things this looks good to me.

**Comment by Archmonger** [comment]
Don't think I'll be able to tackle those changes until the 4th. On vacation, I'm not in my home state right now.

File: src/django_idom/websocket_consumer.py (hunk 1 above)
Comment (Archmonger) [reply src/django_idom/websocket_consumer.py:24]: Disconnect is a function we have within Django IDOM's consumer. So it'll never be none.

**Comment by rmorshea** [comment]
No worries.

File: src/django_idom/websocket_consumer.py (hunk 1 above)
Comment (rmorshea) [reply src/django_idom/websocket_consumer.py:24]: What `Callable[[], None]` indicates is that the intended usage is `ws_conn.disconnect()` where the return value of `disconnect()` is `None`.

Given that it comes directly from Django IDOM's consumer, the signature would be `Callable[[int], Awaitable[None]]` with the intended use `await ws_conn.disconnect(code)`. This is fine, but we should probably have a sync version of this function too called `disconnect_soon`. This would be a a simple wrapper around `disconnect` that uses `asyncio.ensure_future`.

The interface here should then be:

```python
@dataclass
class WebSocketConnection:
    scope: dict
    disconnect: Callable[[int], typing.Awaitable[None]]
    disconnect_soon: Callable[[int], asyncio.Future[None]]
```

The implementation for `disconnect_soon` ought to look something like:

```python
def disconnect_soon(code: int) -> asyncio.Future[None]:
    return asyncio.ensure_future(disconnect(code))
```

**Comment by Archmonger** [comment]
- Replaced WebSocketConnection `disconnect` with `close`, since close ultimately results in `disconnect` being called. 
- Added `view_id` to WebSocketConnection
- Fixed index.js spacing

File: src/django_idom/websocket_consumer.py (hunk 1 above)
Comment (Archmonger) [reply src/django_idom/websocket_consumer.py]: I don't think a sync helper will be needed since we'll always prefer calling async within our consumer.

Comment (rmorshea) [reply src/django_idom/websocket_consumer.py]: You won't be able to call the async function from within a component directly though since that's a sync function. You'd need to do so in an event handler of component effect hook.

Comment (Archmonger) [reply src/django_idom/websocket_consumer.py]: The django channels recommended way of using async within a sync context is using their conversion helper. We could simply recommend using it to minimize function wrappers.

```python 
from channels.db import database_sync_to_async, sync_to_async
```

---

With that in mind, has there been considerations for async components within IDOM core? It's theoretically possible for someone to have a long polling component action. Would be best if that doesn't lock up IDOM and Django Channels.

For example within Conreq I allow most things to be either via inspect.

```python
from inspect import iscoroutinefunction

async def sign_up(request):
    view = app.config.sign_up_view
    if iscoroutinefunction(view):
        return await view(request)
    return view(request)
```

File: src/django_idom/websocket_consumer.py [hunk 3]
```diff
@@ -15,14 +18,37 @@
 _logger = logging.getLogger(__name__)
 
 
-class IdomAsyncWebSocketConsumer(AsyncJsonWebsocketConsumer):
+@dataclass
+class WebsocketConnection:
+    scope: dict
+    close: Callable[[Optional[int]], Coroutine[Any, Any, None]]
```
Comment (rmorshea) [inline src/django_idom/websocket_consumer.py:24]: You should be able to just replace `Coroutine[Any, Any, None]` with `Awaitable[None]`

Comment (Archmonger) [reply src/django_idom/websocket_consumer.py]: Alright, completed.

File: src/django_idom/websocket_consumer.py (hunk 1 above)
Comment (rmorshea) [reply src/django_idom/websocket_consumer.py]: So in this case I don't think those helpers from Django apply here since what we'd want is one that converts an async function to a sync one which is what `asyncio.ensure_future` is meant to do.

As for async components, in the past they actually were async. However there are two reasons they aren't now:

1. **Rendering the layout is a serial task** - technically you could render sibling components in parallel, but in 99% of cases that has no benefit at the cost of making things more complicated for the author to understand.
2. **You can use async functions in the [`use_effect`](https://idom-docs.herokuapp.com/docs/life-cycle-hooks.html#async-effects) hook** - an action like triggering the websocket to disconnect is a perfect example of a "side effect" that belongs withing the `use_effect` hook.

Comment (Archmonger) [reply src/django_idom/websocket_consumer.py]: Apologies for the late reply.

Django ASGI-REF has the inverse of the previously mentioned function. For example: [async_to_sync](https://docs.djangoproject.com/en/3.2/topics/async/#asgiref.sync.sync_to_async). I don't think the developer manually disconnecting the websocket is common enough scenario to justify overcomplicating our function helpers. Frankly the only scenario I can think of it being utilizing is force disconnecting the WS after failed user authentication.

---

The discussion about async within IDOM should probably be moved to another thread but I'll quickly post my thoughts.

I agree that in most cases there isn't much benefit to async components, and they add complexity as you mentioned. That's pretty much the same conclusion that the Django team reached with async views. However, I think long term supporting both is ideal in order to prevent blocking/polling tasks from stalling IDOM.

The solution the Django team took was to the developer to decide whether to use sync or async views. To my knowledge, sync views are run in threads, and async views are handled through an event loop.

**Comment by Archmonger** [comment]
As per my previous comment on your review, let me know what additional changes are needed in this PR.

**Comment by rmorshea** [comment]
Been really busy lately. Will try to get to this next week.

**Comment by Archmonger** [comment]
I updated the PR description.

Let me know if you need me to split this PR into separate parts to be more easily digestible.

File: tests/test_app/templates/base.html [hunk 4]
```diff
@@ -1,23 +1,22 @@
...
+<body>
+  <h1>IDOM Test Page</h1>
+  <div>{% idom_component "test_app.components.HelloWorld" class="hello-world" %}</div>
+  <div>{% idom_component "test_app.components.Button" class="button" %}</div>
+  <div>{% idom_component "test_app.components.ParametrizedComponent" class="parametarized-component" x=123 y=456 %}
+  </div>
+  <div>{% idom_component "test_app.components.SimpleBarChart" class="simple-bar-chart" %}</div>
+</body>
+
+</html>
```
Comment (rmorshea) [inline tests/test_app/templates/base.html:22]: missing end of file new line

File: src/django_idom/config.py [hunk 5]
```diff
@@ -10,6 +10,7 @@
 IDOM_BASE_URL = getattr(settings, "IDOM_BASE_URL", "_idom/")
 IDOM_WEBSOCKET_URL = IDOM_BASE_URL + "websocket/"
 IDOM_WEB_MODULES_URL = IDOM_BASE_URL + "web_module/"
+IDOM_WS_RECONNECT_TIMEOUT = getattr(settings, "IDOM_WS_RECONNECT_TIMEOUT", 604800)
```
Comment (rmorshea) [inline src/django_idom/config.py:13]: So I need to rename this parameter in the JS client. This actually defines the ceiling for the delay between reconnection attempts. The client doesn't actually have a way to stop reconnection attempts at the moment :/

Maybe `IDOM_WS_MAX_RECONNECT_DELAY` would better communicate this.

**Review by rmorshea — CHANGES_REQUESTED** [review]
LGTM - just minor nits.

Once I implement the `use_context` hook I think we'll be able to get rid of this websocket parameter and just allow the user to access the socket from anywhere:

```python
@component
def MyComponent():
    socket = use_context(WebSocketContext)
    ...
```

With everything else that needs to happen that's probably a ways off though so this should be fine for the time being.

File: tests/test_app/templates/base.html (hunk 4 above)
Comment (Archmonger) [reply tests/test_app/templates/base.html:22]: On the note of EOF new lines, I'm going to make a PR to both DJI and Core to add a vscode configuration to automate things like this for new devs passing by.

File: src/django_idom/config.py (hunk 5 above)
Comment (Archmonger) [reply src/django_idom/config.py:13]: I'll also PR IDOM core to get a true "timeout".

Comment (rmorshea) [reply src/django_idom/config.py:13]: I'd actually like to modify how reconnections happen so its more configurable. It should accept a callback that receives the last delay, and the total elapsed time as its arguments and returns the next delay.

**Comment by Archmonger** [comment]
Side note...

We should also consider adding `pre-commit.ci` to all idom-team repos. It's an extremely handy utility for automating common code cleanup during the PR stage.

Since both of us use `black` on our repos, you can just take my fairly extensive [pre-commit.ci configuration file](https://github.com/Archmonger/Conreq/blob/a11f5748e9a41646bcce35fb70c259c6cac61905/.pre-commit-config.yaml).

**Comment by Archmonger** [comment]
All comments have been resolved.

**Comment by rmorshea** [comment]
> I see that IDOM Core is using [pre-commit] currently, but it's using a fairly minimal config.

From my experience, pre-commit has been most helpful in resolving issues that can be fixed automatically (e.g. black, sorting imports) which is why I've kept it fairly minimal. Linters like `flake8` and `pylint` that just report errors can be annoying if you want to introduce a commit that represents intermediate progress. So, including pre-commit hooks that add EOF new lines and strip out extra white space would be great to have though.

A longer-term solution to this would be to make an `idom-team/python-package-template` repo that pre-bakes all the standards we want to enforce.

**Review by rmorshea — APPROVED** [review]

**Comment by Archmonger** [comment]
> A longer-term solution to this would be to make an `idom-team/python-package-template` repo that pre-bakes all the standards we want to enforce.

I recommend using the [repo-file-sync](https://github.com/BetaHuhn/repo-file-sync-action) GitHub action to automate standardized configurations across `idom-team` repos.

That's what we use at [Jazzband](https://github.com/jazzband/) alongside our [template repository](https://github.com/jazzband/.github).
